File: sorting/insertion/insertion.cpp
#include "../../shared/includes/helpers.cpp"

void insertion_sort(int *array, int size) {
  int key, search_index;
  for (int index = 1; index < size; index++) {
    key = array[index];
    search_index = index - 1;
    while (search_index >= 0 && array[search_index] > key) {  
      array[search_index+1] = array[search_index];
      search_index--;
    }
    array[search_index+1] = key;
  }                
}

int main() {
  int ar[] = {5,1,4,3,2};
  int size = 5;
  cout << "Before sort:" << endl;
  printArray(ar, size);
  insertion_sort(ar, size);
  cout << "After sort:" << endl;
  printArray(ar, size);
}

File: sorting/insertion/insertion.rb
def insertion_sort(array)
  for index in 1..(array.length-1)
    key = array[index]
    search_index = index - 1
    while search_index >= 0 and array[search_index] > key
      array[search_index+1] = array[search_index]
      search_index = search_index - 1
    end
    array[search_index+1] = key
  end
  array
end

arr = [5,2,1,3,4]
puts "Before sort:"
puts arr.inspect
insertion_sort(arr)
puts "After sort:"
puts arr.inspect